shotover/src/transforms/kafka/sink_cluster/mod.rs

```diff
@@ -462,6 +499,57 @@ impl KafkaSinkCluster {
                     ..
                 })) => results.push(self.route_to_controller(message).await?),
 
+                Some(Frame::Kafka(KafkaFrame::Request {
+                    body: RequestBody::ApiVersions(_),
+                    ..
+                })) => {
+                    let connection = self
+                        .nodes
+                        .get_mut(0)
+                        .unwrap()
+                        .get_connection(&self.connection_factory)
+                        .await?;
+
+                    let (tx, rx) = oneshot::channel();
+                    connection
+                        .send(Request {
+                            message: message.clone(),
+                            return_chan: Some(tx),
+                        })
+                        .map_err(|_| anyhow!("Failed to send"))?;
+
+                    results.push(rx);
+                }
+
+                Some(Frame::Kafka(KafkaFrame::Request {
+                    body: RequestBody::SaslHandshake(_),
+                    ..
+                })) => {
+                    let (tx, rx) = oneshot::channel();
+                    self.route_to_first_node(message.clone(), Some(tx)).await?;
+
+                    self.connection_factory
+                        .add_handshake_message(message.clone());
+
+                    results.push(rx);
+                }
+
+                Some(Frame::Kafka(KafkaFrame::Request {
+                    body: RequestBody::SaslAuthenticate(_),
+                    ..
+                })) => {
+                    let (tx, rx) = oneshot::channel();
+                    self.route_to_first_node(message.clone(), Some(tx)).await?;
+                    self.connection_factory.add_auth_message(message.clone());
+                    results.push(rx);
+                    self.sasl_status.set_handshake_complete();
+
+                    // now that we have the full handshake, open a connection to all nodes
+                    for node in &mut self.nodes {
```

Comment:
This linearly opens a connection to every node. This will be quite slow in large connections.
But I think we are better off just not doing this at all, so that the cluster sink will lazily open connections as they are needed.